### PR TITLE
Backend for Dockstore Dictionary

### PR DESCRIPTION
**Pull request by aofarrel**
This is the big brother of #200. Instead of focusing on the content of the dictionary, this focuses on how it's made.
* Generates the glossary dynamically based on entries in `/_attic/glossary_entries.py`
* Only adds one dependency ([glossarpy](https://github.com/aofarrel/glossarpy))
  * Added to requirements.txt, so pip install -r command remains the same
  *  Tested installation in a fresh venv
* Pythonic! 🐍 
* Nestles itself into the existing Makefile with a few changes as possible
  * RST glossary + plaintext list of every defined word within it can be rendered on its own via `make glossary`
  * `make html`, `make singlehtml`, etc will also generate the glossary; you do not need to run `make glossary` first
  *  preserves existing behavior wherein `make literally_anything_except_glossary` getting passed to Sphinx
  * Sphinx's kind of weird handling of `make help` untouched
  * SPHINXOPTS, SPHINXBUILD, etc untouched
  * make.bat untouched
* Keeps a running list of what entries are included in glossary_entries_list_dynamic.txt (but I can remove that if it's confusing)

Notes for reviewers:
* Any differences involving dictionary.rst or glossary_entries.py should be ignored when reviewing this PR, as both are dependent upon output I receive from #200.
* I'm aware of the linkcheck issue, will be fixed Friday (was eager to get this out before I'm OOO Thursday)

### CODE REVIEW

**Review by coverbeck — COMMENTED** [review]
I did a `make html`, and `dictionary.rst` became dirty, because I ran it on a more recent date that the committed version.

It would be nice if it didn't do that, otherwise we're going to have to be committing the file every time we do a PR.

Ideas:

1. .gitignore dictionary.rst
2. Don't put the update date in there

I lean towards one; generally avoid committing generated files, unless there's some external dependency on them.

Added @david4096 as a reviewer since he's a lot more familiar with Python than I am.

File: requirements.txt [hunk 1]
```diff
@@ -3,3 +3,4 @@ myst-parser==0.16.1
 sphinx_rtd_theme==1.0.0
 git+https://github.com/pdavide/sphinxcontrib-discourse
 attrs==21.4.0
+glossarpy>=0.0.6
```
[the remark this answers is not in the document]
Comment (aofarrel) [reply requirements.txt]: Fixed in https://github.com/dockstore/dockstore-documentation/pull/201/commits/d211202f5df9c5b062e24f3c2df4d8d6ba697f4f

File: docs/_attic/glossary_entries_list_dynamic.txt [hunk 2]
```diff
@@ -0,0 +1,77 @@
+.dockstore.yml
```
[the remark this answers is not in the document]
Comment (aofarrel) [reply docs/_attic/glossary_entries_list_dynamic.txt:1]: This is auto-generated too. It's just a reference, yes. I can remove it and block its generation if you think it's confusing.

**Comment by aofarrel** [comment]
> I did a `make html`, and `dictionary.rst` became dirty, because I ran it on a more recent date that the committed version.
> 
> It would be nice if it didn't do that, otherwise we're going to have to be committing the file every time we do a PR.
> 
> Ideas:
> 
> 1. .gitignore dictionary.rst
> 2. Don't put the update date in there
> 
> I lean towards one; generally avoid committing generated files, unless there's some external dependency on them.

The index page depends on dictionary.rst due to how the sidebar works in Sphinx, so I think 2 is our best option and made the changes necessary for that in https://github.com/dockstore/dockstore-documentation/pull/201/commits/d211202f5df9c5b062e24f3c2df4d8d6ba697f4f. Any further `make whatever`s will not put the update date in there.

File: docs/dictionary.rst [hunk 3]
```diff
@@ -1,5 +1,4 @@
-.. This file was created using GreatGloss. It is highly recommended to update the
-		source file that this page was generated from rather than modifying it directly.
+.. DO NOT EDIT THIS FILE. Update the Python source file that this page was generated from instead.
```
[the remark this answers is not in the document]
Comment (aofarrel) [reply docs/dictionary.rst]: Fixed in https://github.com/dockstore/dockstore-documentation/pull/201/commits/7f121ac7f8b8aad083f07cf3d009c0f88bbcd88b and https://github.com/aofarrel/glossarpy/commit/c419b003b16ffc35b730d151686b9fa9b7bcb1bf

**Comment by aofarrel** [comment]
Wikipedia is timing out now, probably sent them too many requests in a short time... I'll keep an eye on linkcheck.

**Review by denis-yuen — APPROVED** [review]

File: docs/_attic/glossary_entries_list_dynamic.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (aofarrel) [reply docs/_attic/glossary_entries_list_dynamic.txt:2]: Does https://github.com/dockstore/dockstore-documentation/pull/201/commits/8c9ef6ac9b19d7643503940061544289cce00741#diff-b541cd3f5a84ff8bbbe0b5746d9d565103eb673a33af19da9b1f5c340d576322 address what you meant or do you want a more specific explanation of usage?

**Comment by aofarrel** [comment]
David requested a breakdown of the difference between this and Sphinx's built-in glossary directive.

Similarities:
* both require the user install Python to render HTML files
* both indent the entire definition paragraph
* both can have their entries programmatically sorted (although Sphinx's built-in does not have the ability to generate a simple list of all entries's names, which is something glossarpy can do)
* both create internal links that serve as anchors that can be used elsewhere in documentation
* both will have Sphinx correctly throw warnings if invalid links (internal or external) exist during `make html`
* both are compatible with `make linkcheck`
* glossarpy has no dependencies and is fully controlled by me (would be willing to transfer to Dockstore/TLP like Brian H did herzog) and won't break unless I break it -- this is better than most of our other dependencies (see the mysterious disappearance of attrs I had to debug a few weeks ago)

Differences:
* Words defined by glossarpy are internally considered RST headers, while words defined by Sphinx-glossary are not considered RST headers. This allows glossarpy's entries to be indexed individually elsewhere and appear on the sidebar (although I am considering changing our sidebar to not show headers at all). 
   * If it turns out we **_don't_** want RST headers, it would be trivial for me to modify glossarpy to create non-header bold text instead of headers
   * If it turns out we **_do_** want RST headers, it would not be possible to continue using Sphinx-glossary (although a mass-change to a different format could be done with some Sublime text wizardry)
* glossarpy can generate [a list containing the names of all entries](https://github.com/dockstore/dockstore-documentation/pull/201/files#diff-b541cd3f5a84ff8bbbe0b5746d9d565103eb673a33af19da9b1f5c340d576322) for quick reference by maintainers looking to see what terms have already been defined or not defined. Sphinx-glossary cannot do this, as far as I'm aware.
* The input file that glossarpy reads has simpler syntax than the RST file it outputs. A Sphinx-glossary-styled RST document would still be more complicated than a glossarpy input (unless all formatting would be removed, but even then, many things glossarpy handles automatically would still have to be handled by hand).
* glossarpy handles all of the annoying common syntax pitfalls of Sphinx which frequently cause warnings/errors automatically, while using just glossary-Sphinx would require handling those manually. These include:
  * number of characters under a heading line
  * consistent heading style
  * differing requirements for newlines for different formatting directives
  * nested formatting
  * internal links
* glossarpy will make different parts of a defined term (the pronunciation, the links to similar terms, the external links, etc) stand out by changing up the formatting. For example, with glossarpy, an abbreviation always starts with "abbreviation of" in italics. Some of that formatting can be done manually using Sphinx-glossary styled RST, but the maintainer would have to remember to be consistent with that formatting across dozens of terms, and there isn't an easy way to check for that consistency except by rendering the page manually and visually checking every aspect of every definition.
* A glossarpy user referencing another glossary term only needs to know to enclose the term they want to reference in brackes in a string, ie "A federated cloud platform funded by [NHGRI]." A Sphinx-glossary user must use the :term: syntax, which is used exclusive for Sphinx-glossaries and does not work with the more commonly used :ref: syntax for internal cross-referencing.
* glossarpy will automatically generate a TOC that takes up x amount of columns (default is 3). Sphinx can automatically generate a local TOC, but it will be one-column, which is a lot of space wasted and extra scrolling needed for an 80+ entry glossary. Sphinx users who are aware of the `hlist` directive can also generate what is essentially a local TOC with x number of columns, but they would have to manually list all entries by hand, and would have to manually maintain that list in correct alphabetical order as entries are added/removed.

**Review by david4096 — APPROVED** [review]
Thanks for the explanation and adding the documentation on how to maintain it.

**Pull request merged**